Add pretty printing for `impl Display for {Query, Value Array, Object}`.

finnbear opened the pull request:
Thank you for submitting this pull request! We appreciate you spending the time to work on these changes.

## What is the motivation?

The current `impl Display for Query` produces output like:
```rust
let query = parse("SELECT * FROM {foo: [1, 2, 3]};").unwrap();

println!("{}", query);

SELECT * FROM { foo: [1, 2, 3] };
```

## What does this change do?

Adds support for [alternate](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.alternate) formatting (currently just for for `Query`) to produce output like:
```rust
println!("{:#}", query);

SELECT * FROM {
	foo: [
		1,
		2,
		3
	]
};
```

## Unanswered questions

- Should pretty printing by the #/alternate flag work on other types (e.g. `println!("{:#}", Value | Object | Array | etc.)`)? With the current approach, each `impl Display` that needs to handle the `#`/alternate flag needs a tiny extra bit of code.

## TODO

- [x] Add documentation
- [x] Fix a bug causing extraneous pretty printing on every comma separated sequence
- [x] Cleanup

## What is your testing strategy?

Added a test for the above case.

## Is this related to any issues?

Not to my knowledge but it is related to Discord DM's :joy:

## Have you read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)?

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

finnbear commented on lib/src/sql/statements/ifelse.rs:
```diff
@@ -48,21 +50,21 @@ impl IfelseStatement {
+		Display::fmt(
+			&Fmt::new(
+				self.exprs.iter().map(|args| {
+					Fmt::new(args, |(cond, then), f| write!(f, "IF {} THEN {}", cond, then))
+				}),
+				fmt_separated_by(" ELSE "),
+			),
 			f,
-			"{}",
-			self.exprs
-				.iter()
-				.map(|(ref cond, ref then)| format!("IF {} THEN {}", cond, then))
-				.collect::<Vec<_>>()
-				.join(" ELSE ")
 		)?;
 		if let Some(ref v) = self.close {
 			write!(f, " ELSE {}", v)?
 		}
-		write!(f, " END")?;
+		f.write_str(" END")?;
 		Ok(())
 	}
 }
```
From now on, using `format!` in `impl Display` is not only inefficient but incorrect. Instead of passing through formatting actions, allowing the pretty printer to add newlines and indentation in between them, it buffers up formatting actions into a `String`.

The solution is to use `sql::fmt::Fmt` along side a formatting function. Since a common formatting activity is joining an iterator of `impl Display` by some other `impl Display`, I made a helper called `fmt_separated_by` that produces a closure capable of doing that. It's like a general version of `Fmt::[pretty_]comma_separated`.

tobiemh commented:
Hey @finnbear there are two things I can think of with this...

1. I need to be able to output (with spacing) any SurrealQL type (`Value`, `Object`, `Array`). It can't always be guaranteed to be a `Query`.
2. I would like to be able to customise other types of statement in an alternate way (not generic). Therefore it would be great to be able to customise each implementation of the pretty-printing logic separately. For instance a `DEFINE TABLE` statement might be written like this in pretty-print mode...
```sql
DEFINE TABLE test SCHEMAFULL
    PERMISSIONS
        FOR create, update, delete NONE
        FOR select
            WHERE public = true
;
```

finnbear commented:
> I need to be able to output (with spacing) any SurrealQL type (Value, Object, Array). It can't always be guaranteed to be a Query.

Ok, I'll add support for starting with other SurrealQL types :ok_hand: 

> 2\. would like to be able to customise other types of statement in an alternate way (not generic).

Are you implying that the constituent parts of a statement (e.g. the `SET` clause a.k.a. `surrealdb::sql::data::Data::SetExpression`) might be pretty-printed differently depending on which statement they are part of?

```sql
CREATE thing
   SET -- Default rule: SET within CREATE will use one line per assignment
      a = 5,
      b = {
          c: 6
      }
;

UPDATE thing
    SET a = 5, b = { -- Custom rule: SET within UPDATE will use one line total
        c: 6 -- But the contents are still pretty-printed
    }
;
```

Or are you just saying that the pretty printing should be able to produce the specific output you want for the top-level clauses within a particular statement (e.g. `SCHEMAFULL` is inline, `PERMISSIONS` is indented, `etc.):
```
DEFINE TABLE test SCHEMAFULL
    PERMISSIONS
        FOR create, update, delete NONE
        FOR select
            WHERE public = true
;
```
The former would require tracking some additional state like which statement type it is currently formatting. The latter is considerably easier.

tobiemh commented:
Just the latter! Basically for a `DEFINE TABLE` statement, if it has a `PERMISSIONS` clause, then add a new line break, indent, and then hand off to the `PERMISSIONS` fmt implementation.

finnbear commented:
@tobieh I've just customized `DEFINE TABLE` and `PERMISSIONS` to do pretty much* what you described, and added a test** :tada:

*the permissions are aggregated in the order select, create, update, delete as opposed to create, update, delete, select.

**The tests all pass but fail to compile in GitHub actions due to a linker issue that is likely unrelated to this PR because it also affected [another PR](https://github.com/surrealdb/surrealdb/actions/runs/3841478076/jobs/6541741609), at least until that other PR removed cargo caching support for GitHub actions.

I also added support to directly pretty print `Value`, `Array`, and `Object`. It's not clear whether you also want support for directly pretty printing `Algorithm`, `Base`, `Comment`, `Common`, `Cond`, `Constant`, `Data`, `Datetime`, `Dir`, `Duration`, `Edges`, `Ending`, `Escape`, `Expression`, `Fetch`, `Field`, `Function`, `Geometry`, `Graph`, `Group`, `Id`, `Ident`, `Idiom`, `Kind`, `Limit`, `Model`, `Number`, `Object`, `Operation`, `Operator`, `Order`, `Param`, `Part`, `Paths`, `Permission`, `Query`, `Range`, `Regex`, `Script`,`Split`, `Start`, `Statement` (including all variants), `Strand`, `Subquery` (including all variants), `Table`, `Thing`, `Timeout`, `Uuid`, `Version`, and/or `View` but it would certainly be possible (either in this PR or future PRs).
